server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/LocalShardsBalancer.java

```diff
@@ -931,11 +936,26 @@ AllocateUnassignedDecision decideAllocateUnassigned(final ShardRouting shard) {
                         final int minNodeHigh = minNode.highestPrimary(shard.getIndexName());
                         updateMinNode = ((((nodeHigh > repId && minNodeHigh > repId) || (nodeHigh < repId && minNodeHigh < repId))
                             && (nodeHigh < minNodeHigh)) || (nodeHigh > repId && minNodeHigh < repId));
+                        minNodes.add(node);
                     } else {
                         updateMinNode = currentDecision.type() == Decision.Type.YES;
+                        /* If updateMinNode is true, it means the earlier nodes had decision type THROTTLE. We will need to clear the list,
+                         * and add new nodes to the list.
+                         */
+                        if (updateMinNode) {
+                            minNodes.clear();
+                            minNodes.add(node);
+                        }
                     }
                 } else {
                     updateMinNode = currentWeight < minWeight;
+                    /* Since we have found nodes with less weight. We will need to clear the earlier minNodes list
+                     * and add the new nodes to the list.
+                     */
+                    if (updateMinNode) {
+                        minNodes.clear();
+                        minNodes.add(node);
+                    }
```

Comment:
As discussed, lets not mix up new logic with the existing logic. Lets try to have an abstraction that can be used for both the approaches and we can overload or have 2 methods for getting the minNodes, weight, decision.

Reply:
Updated.